databricks_cli/pipelines/api.py

```diff
@@ -0,0 +1,187 @@
+# Databricks CLI
+# Copyright 2017 Databricks, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"), except
+# that the use of services to which certain application programming
+# interfaces (each, an "API") connect requires that the user first obtain
+# a license for the use of the APIs from Databricks, Inc. ("Databricks"),
+# by creating an account at www.databricks.com and agreeing to either (a)
+# the Community Edition Terms of Service, (b) the Databricks Terms of
+# Service, or (c) another written agreement between Licensee and Databricks
+# for the use of the APIs.
+#
+# You may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from hashlib import sha1
+import os
+
+from six.moves import urllib
+
+from databricks_cli.sdk import DeltaPipelinesService
+from databricks_cli.dbfs.api import DbfsApi
+from databricks_cli.dbfs.dbfs_path import DbfsPath
+
+
+# These imports are specific to the credentials part
+from databricks_cli.configure.config import get_profile_from_context
+from databricks_cli.configure.provider import get_config, ProfileConfigProvider
+from databricks_cli.utils import InvalidConfigurationError
+
+BUFFER_SIZE = 1024 * 64
+base_pipelines_dir = 'dbfs:/pipelines/code/'
+
+
+class PipelinesApi(object):
+    def __init__(self, api_client):
+        self.client = DeltaPipelinesService(api_client)
+        self.dbfs_client = DbfsApi(api_client)
+
+    def deploy(self, spec, headers=None):
```

Comment:
Is there a place where we document where this spec might look like?

Reply:
There is a link to an example spec in the doc https://docs.google.com/document/d/1K3iaOzfv0OvIsvZsdqiFDfvVFMh8Oh58wA2x0P8rtHQ/edit#heading=h.twb6crp48eby

Reply:
I think it'd be very valuable to document it in this repo as well: in code as a comment and as an example somewhere as well.

Reply:
I don't think we should put an example spec here, since the spec may change in the future and this function is (mostly) forwarding the spec to our delta pipelines API. The spec specification and examples should live on the docs site (and we'll add a link to that from here when we get that up). 

I do think we should add documentation here saying:
 - What a spec is (a Delta Pipelines deployment specification that explains how to run a Delta Pipeline on Databricks)
 - That it gets forwarded to the Delta Pipelines deployment API
 - What we actually do check for and do with it in this function (look for all its libraries that are specified as local paths and upload them to DBFS).

Reply:
I've added this documentation in pipelines/cli.py deploy